Fix Various MapShed Issues

rajadain opened the pull request:
## Overview

Fixes the following issues:

 * Maps NLCD 21 to Low Density Mixed instead of Open Land
 * Looks up C and P values which were previously missing
 * Corrects day length calculation caused by an index mismatch
 * Corrects LS calculation which was previously mixing NHD stream percentages with DRB stream lengths
 * Corrects KV calculation which was previously using erosion coefficients instead of evapotranspiration coefficients

## Testing Instructions

Checkout the branch. Debug Celery

    ./scripts/debugcelery.sh

Select the **LIttle Neshaminy Creek** HUC12 right above Philadelphia as an area of interest.

![image](https://cloud.githubusercontent.com/assets/1430060/15948610/b59033c4-2e6f-11e6-9c78-799dd19ef347.png)

Create a MapShed project and export "Current Conditions" as a GMS file.

Move the GMS file to a Windows VM. Open the GMS file using GWLF-E 1.4.0 which can be found at smb://fileshare/projects/Stroud_ModelMyWatershed/downloads/mapshed/GWLF-E-1.4.0.exe.

Compare the results of Transport Data and Nutrient Data with that of [`input_365.gms`](https://github.com/WikiWatershed/model-my-watershed/files/308010/input_365.gms.txt). The values will not be exact, since the data sources are different, but they should be within each others ballpark.

Connects #1357

rajadain commented:
Here are my comparisons:

Little Neshaminy Transport Data
![little_neshaminy_kvfix_forreal_td](https://cloud.githubusercontent.com/assets/1430060/15948786/c8777b22-2e70-11e6-8da9-f356cbe839b2.jpg)

BME Transport Data
![input_365_td](https://cloud.githubusercontent.com/assets/1430060/15948797/d8bc1c5e-2e70-11e6-9494-638989976093.jpg)

Little Neshaminy Nutrient Data
![little_neshaminy_kvfix_forreal_nd](https://cloud.githubusercontent.com/assets/1430060/15948806/e2a8c596-2e70-11e6-9a83-9e65663727ac.jpg)

BME Nutrient Data
![input_365_nd](https://cloud.githubusercontent.com/assets/1430060/15948811/eb51b554-2e70-11e6-9e64-0e4ea47141dd.jpg)

caseypt commented on src/mmw/apps/modeling/mapshed/tasks.py:
```diff
@@ -252,6 +256,29 @@ def nlcd_streams(sjs_result):
+    result = parse_sjs_result(sjs_result)
+    total = sum(result.values())
+
+    lu_stream_pct_drb = [0.0] * 16
```
Consider adding a comment to describe the significance of `16`.

rajadain replied:
It's just the total number of land use types, but I will add a comment / variable to make that more clear.

caseypt commented on src/mmw/apps/modeling/mapshed/tasks.py:
```diff
@@ -252,6 +256,29 @@ def nlcd_streams(sjs_result):
+    for nlcd, stream_count in result.iteritems():
+        lu = get_lu_index(nlcd)
+        if lu is not None:
+            lu_stream_pct_drb[lu] += float(stream_count) / total
```
Should we protect against a divide by zero error here?

rajadain replied:
Good catch!

rajadain replied:
Although we probably don't need to, because if `total` has a value it is because some item in result has a value. Thus, when `total` is 0, we simply won't enter the `for nlcd, stream_count in result.iteritems()` loop.

caseypt replied:
Is it possible for `result` to look like `{ nlcd_value: 0, nlcd_value: 0, nlcd_value: 0, etc.. }`?

rajadain replied:
No, the geoprocessing service does not have any idea of what the "full set" of NLCD values is like. In its response it only reports the cell values it finds. Thus, if a certain polygon has only 3 types of land uses, the output will also only have 3.

caseypt replied:
Got it.

caseypt commented on src/mmw/apps/modeling/mapshed/tasks.py:
```diff
@@ -437,6 +464,12 @@ def geop_tasks(geom, errback):
+        definitions.append(('nlcd_streams',
+                            {'polygon': [geom.geojson],
+                             'vector': streams(geom, drb=True)},
+                            nlcd_streams_drb))
```
If an AoI is within the DRB, it looks like the `nlcd_streams` task will be run twice; once in the beginning, and once at the end.

rajadain replied:
That's correct. While we prefer to use the DRB streams for calculating LS if they are available, we _always_ use the NHD streams for Stream Bank Erosion, which are calculated here: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/apps/modeling/mapshed/tasks.py#L137-L142. Thus, we need the NHD streams in all cases, and the DRB streams only in the cases when they are available.

caseypt replied:
Got it. Is it okay that the `opname`s are the same?

rajadain replied:
That's a bit confusing, probably warrants a comment, but yes. That actually points to the JSON block that is sent to the geoprocessing service, which is defined here: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L386-L399

Both requests use the same JSON block, just different values in the `vector` field.

lewfish commented on src/mmw/apps/modeling/mapshed/calcs.py:
```diff
@@ -575,3 +575,35 @@ def sed_a_factor(landuse_pct_vals, cn, AEU, AvKF, AvSlope):
+        0.8,   # Turf Grass
+        ag_p,  # Open Land
+        ag_p,  # Bare Rock
+        0.0,   # Sandy Areas
```
In `Class1.vb`, it looks like `PFactor(9)` is equal to `P14` which is equal to `Pf1-Pf5`, so Sandy Areas should also be `ag_p`.

caseypt commented on src/mmw/mmw/settings/gwlfe_settings.py:
```diff
@@ -463,6 +463,42 @@
     'SSLDR': 4.4,
     'SSLDM': 2.2,
     'WeatherNull': -99999,  # This value is used to indicate NULL in ms_weather dataset
+    'ETGrowCoeff': [
```
Can you point me to the source of these so I can double check them?

rajadain replied:
Class1.vb@1.3.0:4393-4470

rajadain replied:
Sorry, that's for P Factor.

For ET take a look at Class1.vb:4989-4995 `TransMonthData`, and see where it is called from (sorry don't have the file with me right now)

caseypt replied:
Looks good.

caseypt replied:
Yeah I ended up finding them. Thanks.

lewfish commented on src/mmw/apps/modeling/mapshed/calcs.py:
```diff
@@ -575,3 +575,35 @@ def sed_a_factor(landuse_pct_vals, cn, AEU, AvKF, AvSlope):
+        ag_p,  # Forest
+        0.1,   # Wetland
+        0.1,   # Disturbed
+        0.8,   # Turf Grass
```
In `Class1.vb`, `PFactor(6)` = `P16` which can only ever be 0 or 0.2

lewfish commented on src/mmw/apps/modeling/mapshed/calcs.py:
```diff
@@ -575,3 +575,35 @@ def sed_a_factor(landuse_pct_vals, cn, AEU, AvKF, AvSlope):
+        ag_p,  # Open Land
+        ag_p,  # Bare Rock
+        0.0,   # Sandy Areas
+        1.0,   # Unpaved
```
`PFactor(10)` is 0 or 1 depending on the value of `Lu`. Why do we ignore `Lu` here?

lewfish commented on src/mmw/mmw/settings/gwlfe_settings.py:
```diff
@@ -40,6 +40,8 @@
                 ],
     'Imper': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0,         # Impervious surface area percentage
               0.15, 0.52, 0.87, 0.15, 0.52, 0.87],  # only defined for urban land use types
+    'C': [0.03, 0.42, 0.002, 0.01, 0.1, 0.03, 0.04, 0.001, 0.01, 0.08,  # C Factor Defaults
```
I don't see where C is calculated in the original code.

lewfish replied:
If these numbers are supposed to match C1, C2, .. they appear to be off.

rajadain replied:
They map lines 4393-4436 by land use type

lewfish replied:
I think the 5th value (disturbed) and 10th (unpaved) might be off.  If the comment below is accurate, then disturbed should should be 0.08 and unpaved should be 0.8.
```
        If Lu = 12 Or Lu = 13 Or Lu = 11 Or Lu = 15 Then ' Disturbed, Bare Rock, Unpaved Roads, & Open Land
            C12 = 0.08 : C13 = 0.1 : C11 = 0.8 : C15 = 0.08
```

rajadain replied:
Thanks, I think I used a word document that BME sent. This is more correct, though.

lewfish commented on src/mmw/apps/modeling/mapshed/calcs.py:
```diff
@@ -31,18 +31,20 @@ def day_lengths(geom):
+                                                  ((m + 1) * 30.4375 - 5)))
 
-    return lengths
+    return [round(l, 1) for l in lengths]
```
Looks good.

caseypt commented:
Last two commits look good.

mmcfarland commented:
Corrected reference GMS from BME, with screenshots:

[input_253.gms.txt](https://github.com/WikiWatershed/model-my-watershed/files/309285/input_253.gms.txt)

![screenshot from 2016-06-10 11 44 07](https://cloud.githubusercontent.com/assets/1014341/15970049/ac346e44-2f00-11e6-9e37-ac15c6368704.png)
![screenshot from 2016-06-10 11 43 40](https://cloud.githubusercontent.com/assets/1014341/15970048/ac327a58-2f00-11e6-87a1-1a414755b531.png)

mmcfarland commented:
+1 Screenshots show pretty good tracking of the desktop GIS numbers.  I think there are some outstanding dissolved P discrepancies that won't likely get resolved immediately because they'll require additional datasets that we are holding off implementing.

lewfish commented:
Looks good otherwise.

rajadain commented:
Thanks all. I've addressed all the comments and added a final fix that corrects dissolved phosphorus concentration values:

![untitled_project__current_conditions_nd](https://cloud.githubusercontent.com/assets/1430060/15977424/b4cef6ac-2f26-11e6-97eb-f065b8fbd3fd.jpg)
